feat(coderd): add retention policy configuration

mafredri opened the pull request:
Add `RetentionConfig` with server flags for configuring data retention:

- `--audit-logs-retention`: retention for audit log entries
- `--connection-logs-retention`: retention for connection logs
- `--api-keys-retention`: retention for expired API keys (default 7d)

Note: AI Bridge already has retention flags which are kept separate from `RetentionConfig`, which only affects core `coderd`.

Updates #20743

---

### PR Stack

|  | PR | Title |
|:---:|:---:|:---|
| 👉 | #21021 | feat(coderd): add retention policy configuration |
|  | #21022 | feat(coderd/database/dbpurge): add retention for connection logs |
|  | #21025 | feat(coderd/database/dbpurge): add retention for audit logs |
|  | #21037 | feat(coderd/database/dbpurge): make API keys retention configurable |
|  | #21038 | docs: add data retention documentation |
|  | #21039 | feat: add retention config for `workspace_agent_logs` |

mafredri committed:
feat(coderd): add retention policy configuration

Add `RetentionConfig` with server flags for configuring data retention:

- `--global-retention`: default policy for all retention settings
- `--audit-logs-retention`: retention for audit log entries
- `--connection-logs-retention`: retention for connection logs
- `--api-keys-retention`: retention for expired API keys (default 7d)

Updates #20743

mtojek reviewed (commented):
wait a sec, is it only deployment config, or are you going to use these options soon? if the plan is to continue after some time, maybe park these changes in a side branch. Otherwise the commit message will be confusing.

mtojek reviewed (approved):
Thanks for context 👍 👍

mafredri committed:
advice audit log retention according to docs

mafredri commented on cli/testdata/server-config.yaml.golden:
```diff
@@ -755,3 +755,27 @@ aibridge:
+  # value, or to disable if global is also 0.
+  # (default: 0, type: duration)
+  audit_logs: 0s
+  # How long connection log entries are retained. Set to 0 to use the global
```
Docs say a year so I updated the help to be closer to that 👍🏻

mafredri committed:
chore: remove global retention option

Remove the global retention fallback in favor of explicit per-topic
retention settings. This makes the retention behavior clearer and
easier for operators to reason about.

Each retention setting now:
- Enables retention when set to a non-zero duration
- Disables retention (keep indefinitely) when set to 0

mafredri commented on cli/testdata/coder_server_--help.golden:
```diff
@@ -696,6 +696,35 @@ updating, and deleting workspace resources.
+          How long connection log entries are retained. Set to 0 to use the
+          global retention value, or to disable if global is also 0.
+
+      --global-retention duration, $CODER_GLOBAL_RETENTION (default: 0)
```
Removed for now, we can always intro this as a new feature later, if there is demand 👍🏻
